lxd/storage/drivers/interface.go

```diff
@@ -0,0 +1,52 @@
+package drivers
+
+import (
+	"github.com/lxc/lxd/lxd/migration"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// VolumeType represents a storage volume type.
+type VolumeType string
+
+// VolumeTypeCustom represents a custom storage volume.
+const VolumeTypeCustom = VolumeType("custom")
+
+// VolumeTypeContainer represents a container storage volume.
+const VolumeTypeContainer = VolumeType("containers")
+
+// VolumeTypeImage represents an image storage volume.
+const VolumeTypeImage = VolumeType("images")
+
+// VolumeTypeVM represents a virtual-machine storage volume.
+const VolumeTypeVM = VolumeType("virtual-machines")
+
+// driver is the extended internal interface
+type driver interface {
+	Driver
+
+	create(dbPool *api.StoragePool) error
+}
+
+// Driver repreents a low-level storage driver.
+type Driver interface {
+	// Internal
+	Name() string
+	Version() string
+
+	// Pool
+	Delete(op *operations.Operation) error
+	Mount() (bool, error)
+	Unmount() (bool, error)
+	GetResources() (*api.ResourcesStoragePool, error)
+
+	// Volumes
+	CreateVolume(volType VolumeType, volID int64, volName string, op *operations.Operation) error
+	CopyVolume(srcPool string, srcVolName string, snapshots []string, volType VolumeType, volID int64, volName string, op *operations.Operation) error
+	DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error
+	RenameVolume(volType VolumeType, volName string, newVolName string, op *operations.Operation) error
+
+	// Migration
+	MigrationType() migration.MigrationFSType
+	PreservesInodes() bool
```

Comment:
I think we need an extra `OptimizedImages() bool` symbol as well as a `HasVolume(volType VolumeType, name string) bool` one.

This will then allow a workflow like this from the lxdBackend side during a container creation from an image:
 - OptimizedImages()
   - If true
     - Call HasVolume to see if we have an image volume, note that for ceph & zfs, we'll need that to return `true` for deleted volumes that we kept around for refcounting reasons (`deleted/images/XYZ`).
     - If we don't have an existing volume, call `CreateVolume` for `VolumeTypeImage` and set the `filler` so we can create that missing image volume
     - Then call `CreateVolumeFromCopy` with `srcType` of `VolumeTypeImage` and `srcName` set to the image fingerprint
   - If false
     - Call CreateVolume with a filler set to copy or unpack the image depending on whether filesystem or block

The result is a container created from an image, whether the filesystem has optimized support or not, regardless of whether we're dealing with block or filesystem and dealing with the deleted image trick we have for zfs/ceph and with minimal logic in the Driver and centralized image creation logic in the Backend.

Reply:
@stgraber can you remind me what `OptimizedImages` means, is it when a storage driver caches an unpacked image tarball in a volume with a snapshot so that when creating an instance one can just create a volume from the cached snapshot?

Also, can you explain further about ceph and zfs and the ref counting you referred to? Thanks

Reply:
Correct, OptimizedImages would return true if we're using an image type volume that gets cloned. So this would be true for zfs, lvm, btrfs and ceph, false for dir and cephfs.

Both ZFS and CEPH do not allow the destruction of a volume which is used as the parent of another volume. In both cases, rather than fail the deletion of the image or container because of that refcount, we instead rename the image or container to a name indicating it's been deleted. Then for images, if we're later asked to import that same image again, we can just resurrect the deleted dataset rather than unpacking it again.

Reply:
@stgraber thanks, I've added those new functions.